iTunes: Modularize importer and use `iTunesLibrary` on macOS for compatibility with `Music.app`

fwcd opened the pull request:
This fixes

- #11256
- #9688 (potentially, though a cross-platform solution might still be nice in the future)

### Description

This branch modularizes the iTunes library integration into importers:

- `ITunesMacOSImporter`, using Apple's [`iTunesLibrary`](https://developer.apple.com/documentation/ituneslibrary) framework
    - Used by default on macOS
    - Supports regular and smart playlists, along with (nested) folders
    - Supports both the old and the new database format used by `Music.app`
    - The framework, however, only lets us read the user's default music library, therefore we'll fall back to the XML importer whenever a custom db path is provided on macOS
-  `ITunesXMLImporter`, the current XML parser
    - Used on non-macOS platforms or whenever a custom db path is provided

This modularity should also make it easy to drop in new importers in the future, e.g. a cross-platform parser for the new database format.

### To do

- [x] Modularize iTunes feature
- [x] Migrate current parser to `ITunesXMLImporter`
- [x] Implement `ITunesMacOSImporter`
    - [x] Set up Objective-C++
    - [x] Import tracks
    - [x] Import playlists
        - [x] Support regular and smart playlists
        - [x] Support folders
- [x] Fix the strange macOS CI errors during packaging (see e.g. [this run](https://github.com/mixxxdj/mixxx/actions/runs/4386662711/jobs/7681000385))
- [x] Investigate whether the ad-hoc code signing identity is sufficient (see [this thread on stackoverflow](https://stackoverflow.com/questions/16597284/using-new-ituneslibrary-framework-to-access-itunes-library-on-mac-os-x))
- [x] Investigate whether we need to declare new macOS entitlements (see [this comment on stackoverflow](https://stackoverflow.com/questions/16597284/using-new-ituneslibrary-framework-to-access-itunes-library-on-mac-os-x#comment41021127_16627359))

### Future Directions

- Adapt the folder import logic (ingesting the parent playlist relationships into a multimap and building the `TreeItem` afterwards) for the XML importer
    - This should address issues like
      - #9576
      - #8148
- Read cover art via the `iTunesLibrary` framework too
  - iTunes and `Music.app` in many cases do not store the cover art in the track metadata, making it inaccessible to Mixxx
  - However, there is an API for fetching the cover art via an in-memory `NSImage` from an `ITLibMediaItem`: https://developer.apple.com/documentation/ituneslibrary/itlibmediaitem/1810529-artwork
  - Since cover art isn't stored in the database currently, however, we might either have to
    - modify the schema for this to either store the artwork in the database (possibly space-inefficient for large libraries) or...
    - modify the cover art loader to interface with the `iTunesLibrary` framework (e.g. querying the iTunes ID for a given track), which might be undesirable since this way the iTunes integration would become a leaky abstraction
- Optimize the persistent ID -> database ID mapping
    - We need this since iTunes' persistent IDs use unsigned 64-bit ints, but SQLite only lets us store signed 64-bit ints
    - Currently we use a dictionary and generate a sequence of incrementing indices
        - One advantage is that we get reasonably deterministic indices (0, 1, 2, 3, ...)
        - This requires storing all IDs in memory though (not sure if this is an issue for very large libraries)
    - Potential solutions:
        - Cast the unsigned 64-bit int to a signed 64-bit int for storage in SQLite
            - This would however frequently generate negative indices, which might confuse users and the ORM
        - Perform some form of hashing and investigate how to avoid collisions
            - If we limit ourselves to the positive integers that SQLite can represent, we'd have an output space of 63 bits, which is probably enough, but we'd still want to check the math whether the likelihood of a birthday problem is acceptable
        - Update the database schema
            - Altering the iTunes id column type is a breaking change though, which we'd like to avoid
            - Adding another (unique-constrained) column for persistent IDs [is possible](https://github.com/fwcd/mixxx/commit/fff6b3cddbce38a9970a45f6c7cefc8a81a733fd) without bumping the minimum compatible version, but would require querying the database for every mapping, which would result in a lot of database calls with large playlists

fwcd committed:
ITunesFeature: Factor out ITunesXMLImporter

fwcd committed:
ITunesMacOSImporter: Stub out native iTunes importer for macOS

fwcd committed:
ITunesFeature: Conditionally use ITunesMacOSImporter on macOS

fwcd committed:
ITunesFeature: Extract showOpenDialog

fwcd committed:
ITunesMacOSImporter: Implement media item import

fwcd committed:
ITunesMacOSImporter: Fix imported iTunes durations

fwcd committed:
ITunesMacOSImporter: Remove file

The iTunesLibrary framework only allows us to inspect the default
library, so we might as well reflect that in our API.

fwcd committed:
ITunesFeature: Fall back to XML importer for custom paths

uklotzde commented on src/library/itunes/itunesmacosimporter.h:
```diff
@@ -0,0 +1,20 @@
+#include "library/itunes/itunesimporter.h"
+#include "library/libraryfeature.h"
+
+class ITunesMacOSImporter : public ITunesImporter {
```
How about adopting the official naming by using `AppleMusicApiImporter` instead of inventing new names? Would be much easier to recognize.

https://developer.apple.com/documentation/applemusicapi/

fwcd committed:
ITunesMacOSImporter: Fix style issue

fwcd committed:
ITunesMacOSImporter: Import playlists

fwcd committed:
ITunesMacOSImporter: Use long long for ids

fwcd committed:
ITunesMacOSImporter: Factor out state

fwcd committed:
Factor out ImporterImpl

fwcd committed:
ITunesMacOSImporter: Use autoreleasepool

fwcd committed:
ITunesMacOSImporter: Fix typo

fwcd committed:
ITunesMacOSImporter: Use custom indexing scheme

fwcd committed:
ITunesXMLImporter: Take QString by const ref

fwcd committed:
CMakeLists: Link iTunesLibrary framework weakly

fwcd committed:
ITunesMacOSImporter: Show all audio playlists

fwcd committed:
BundleInstall: Try marking /Library/Frameworks as system libraries

fwcd committed:
ITunesMacOSImporter: Hide some more auto-generated iTunes playlists

fwcd committed:
ITunesMacOSImporter: Split out tree handling

fwcd committed:
ITunesMacOSImporter: Uniquify names

fwcd committed:
ITunesMacOSImporter: Import folder tree

daschuer commented on src/library/itunes/itunesmacosimporter.mm:
```diff
@@ -89,6 +89,17 @@ int persistentIdToDbId(NSNumber* persistentId) {
+            m_playlistDuplicatesByName[name] += 1;
+            return QString("%1 #%2").arg(name).arg(m_playlistDuplicatesByName[name]);
+        } else {
+            m_playlistDuplicatesByName[name] = 1;
```
I don't think we have one duplicate here.
Can we start with zero here, to have a better match with the numbers of duplicates?

daschuer replied:
How can we have duplicates by the way? 
Maybe you can add a brief comment.

fwcd replied:
If the user has multiple playlists with the same name in iTunes then we need to disambiguate them, because the `TreeItem` identifies the playlist by name.

> I don't think we have one duplicate here.

Yeah, the naming is not ideal yet, since we are counting the number of occurrences of a playlist name, not the duplicates, which would be one less (strictly speaking). Perhaps `m_playlistCountsByName` or similar would he better, but I wanted to make it obvious that we were going to use it for duplicate detection.

The idea is, if the user e.g. has three playlists named `Pop` that they are named

- `Pop`
- `Pop #2`
- `Pop #3`

which is similar to what the iTunes XML importer does.

fwcd replied:
I've updated it to count from zero and added an explanatory comment.

daschuer commented on src/library/itunes/itunesmacosimporter.mm:
```diff
@@ -89,6 +89,17 @@ int persistentIdToDbId(NSNumber* persistentId) {
+        auto existing = m_playlistDuplicatesByName.find(name);
+        if (existing != m_playlistDuplicatesByName.end()) {
+            m_playlistDuplicatesByName[name] += 1;
+            return QString("%1 #%2").arg(name).arg(m_playlistDuplicatesByName[name]);
```
```suggestion
            return name + QString()::number(m_playlistDuplicatesByName[name]);
```
performs faster.

fwcd replied:
I think that's missing the space and the hash, i.e. we would get `Pop2` instead of `Pop #2` for example.

fwcd replied:
I'll close this since the existing iTunes XML importer uses a similar disambiguation mechanism and I assume this is an oversight.

fwcd committed:
ITunesMacOSImporter: Use QHash for persistent ID mapping

QHash has less overhead than NSMutableDictionary due to not being
reliant on boxing and message passing.

fwcd committed:
ITunesMacOSImporter: Count the playlist duplicates from 0

fwcd committed:
ITunesMacOSImporter: Add explanatory comment for unique names

Swiftb0y commented:
Do you have a good write-up on Obj-C(++) for C++ developers? Also, can you try to keep the Obj-C code minimal?  introducing more programming languages increases the risk of pulling in code that we rely on but can't maintain. Especially when that programming language is not used outside the apple ecosystem (which we already don't have access to since none of the core devs uses macos).

fwcd commented:
> Also, can you try to keep the Obj-C code minimal?

Yeah, that's something I'm already trying to do, also because C++/Qt containers tend to be quite a bit more performant than ObjC types, the latter rely heavily on boxing.

> Do you have a good write-up on Obj-C(++) for C++ developers?

http://www.jonhoyle.com/maccompanion/columns/200809.html might be worth looking into.

Since we don't use any of the 'fancy' ObjC features here and only what's necessary to interact with the APIs, this should boil down to learning the syntactic differences (only), e.g. `[object method]` instead of `object.method()`. One thing to note is that ObjC objects are always passed as raw pointers, the compiler still performs automatic memory management of them under the hood though (automatic reference counting/ARC to be precise, effectively what a `shared_ptr` does).

fwcd committed:
ITunesMacOSImporter: Format using clang-format

fwcd committed:
ITunesMacOSImporter: Extract qStringFrom(NSString*) function

fwcd committed:
ITunesMacOSImporter: Rename to dbIdFromPersistentId

For consistency with qStringFrom

fwcd committed:
ITunesMacOSImporter: Reformat using clang-format

daschuer commented:
I am unsure what to do with this PR. Is the iTunes integration in 2.3 broken, after a user has received the music.app?

In the unlike case that this PR does any harm, is it possible to disable it and use the XML importer? 

I am considering to merge it to 2.4. Do you think the 3 month beta is good enough for testing. Since I am on Linux, I am personally not able to do a sorrowly test. I think we need a real iTunes user to verify details. 

Any thoughts?

Swiftb0y commented:
> http://www.jonhoyle.com/maccompanion/columns/200809.html might be worth looking into.

Thank you, I'll familiarize and try to review ASAP.

fwcd commented:
> Is the iTunes integration in 2.3 broken, after a user has received the music.app?

Kind of. `Music.app` no longer automatically generates an iTunes XML and keeps it in sync. This means the user might see an outdated view of their library, unless they manually export an XML through the UI, which is still possible through `Music.app`

> In the unlike case that this PR does any harm, is it possible to disable it and use the XML importer?

The user should (even if `Music.app` is installed) always be able to select an XML file in which case we will fall back to the XML importer. The macOS importer is only used when "default library" is selected.

> Do you think the 3 month beta is good enough for testing

Since the API surface that we use is relatively low and we reuse a large part of the existing DB infrastructure, I would say that is sufficient. Ideally some users, especially with large libraries, could test this integration.

fwcd commented:
I've rebased this one on top of 2.4 (too).

Swiftb0y commented:
I found this 70 page document which seems to be fairly comprehensive in addition to the article you linked. We might even want to link to that somewhere in the source or the wiki to make sure people aren't stranded in the future when looking at this: https://pierre.chachatelier.fr/programmation/fichiers/cpp-objc-en.pdf

Swiftb0y commented on src/library/itunes/ituneslocalhosttoken.cpp:
```diff
@@ -0,0 +1,11 @@
+#include "library/itunes/ituneslocalhosttoken.h"
+
+#include <QString>
+
+QString iTunesLocalhostToken() {
+#if defined(__WINDOWS__)
+    return "//localhost/";
+#else
+    return "//localhost";
+#endif
+}
```
this could just be a `static const(expr)` in the header, right?

```cpp
static const QString kiTunesLocalHostToken = QStringLiteral("//localhost"
#if defined(__WINDOWS__)
"/"
#endif
);
```

Swiftb0y commented on src/library/itunes/itunesmacosimporter.mm:
```diff
@@ -0,0 +1,282 @@
+        m_playlistChildsByDbId.insert(parentId, playlistId);
+
+        int i = 0;
+        for (ITLibMediaItem* item in playlist.items) {
```
Makes sense, can you add that rationale as a comment above the loop?
```suggestion
        // We use an Obj-C for-in loop here since the underlying collecting
        // is also an Obj-C type. 
        for (ITLibMediaItem* item in playlist.items) {
```

fwcd replied:
We could do that, though I would _generally_ try to abstain from these kinds of comments that only document syntax (I'm assuming the reader knows basic Objective-C, e.g. we don't document what a C++ for-loop is either).

If the reader sees `for-in` in a `.m` or `.mm` file it should be clear that we are dealing with Objective-C types.

Swiftb0y replied:
I agree in the general case, though I argue that we explicitly can't assume that the reader has knowledge that in depth. Neither the article you linked, nor the 70 page pdf I found mention these pitfalls. As I said, nobody from the mixxx team knows Obj-C (nor do I feel like they have a great desire to learn it). Thats why I would prefer to treat the Obj-C code like language tutorial for people that are looking at this language for the first time. The really basic stuff can be assumed to be known because its the first thing you read in every article (for example invoking selectors) but the intermediate to advanced stuff would preferably be explained.

fwcd replied:
Yeah, but isn't `for-in` pretty self-explanatory?

Swiftb0y replied:
Yes the concept is, but its not self-explanatory why we prefer a Obj-C for-in _over a C++ ranged-based-for_ in this particular case.

fwcd replied:
I've added a quick note on this in 792f1b51bd38aecccdb11864d9928ffb3b0c6559

fwcd commented on src/library/itunes/itunesmacosimporter.mm:
```diff
@@ -0,0 +1,282 @@
+        ITunesImport iTunesImport;
+        iTunesImport.isMusicFolderLocatedAfterTracks = false;
+
+        NSError* error = nil;
+        ITLibrary* library = [[ITLibrary alloc] initWithAPIVersion:@"1.0"
+                                                             error:&error];
+
+        if (library) {
```
Yeah, we probably could. I'm following Apple's example here in which they only check the library object too: https://developer.apple.com/documentation/ituneslibrary/itlibrary/1809375-initwithapiversion?language=objc

It should be safe either way since the documentation mentions that the method is guaranteed to return `nil` if the method errors, but it might be nice to have a descriptive error to present to the user.

Swiftb0y replied:
> It should be safe either way since the documentation mentions that the method is guaranteed to return nil if the method errors, but it might be nice to have a descriptive error to present to the user.

Yes, at least it would be nice to log the issue to console.

fwcd replied:
I've added a `qWarning` that logs the localized description of the error in 88933edcd28bf37d2aade6666481902f7f3399e5

fwcd committed:
ITunesLocalhostToken: Define it as a static constant

fwcd committed:
ITunesFeature: Use atomic bool for cancelImport

fwcd committed:
ITunesMacOSImporter: Add short note on ObjC for-in loops

fwcd committed:
ITunesLocalhostToken: Hopefully fix the MSVC build

Swiftb0y commented on src/library/itunes/ituneslocalhosttoken.h:
```diff
@@ -0,0 +1,10 @@
+
+#include <QString>
+
+static const QString kiTunesLocalhostToken =
+        "//localhost"
+#if defined(__WINDOWS__)
+        "/"
+#endif
+        ;
```
I guess I was too clever for MSVC, then I'd prefer the verbose but still easily readable version. 
```suggestion
#if defined(__WINDOWS__)
static const QString kiTunesLocalhostToken = QStringLiteral("//localhost/");
#elif
static const QString kiTunesLocalhostToken = QStringLiteral("//localhost");
#endif
```

fwcd replied:
Just curious, is there any advantage of explicitly calling `QStringLiteral`? We already declare it as a `QString`, so the explicit conversion feels a bit redundant.

But I'd wait for the MSVC build to complete first, I don't know whether cond-compiles are just problematic within parameter contexts or in expressions in general.

Swiftb0y replied:
Its somewhat redundant because it essentially saves us a single heap-allocation per process IIUC, though it still makes sense IMO.
I think the qt doc explains it best https://doc.qt.io/qt-6/qstring.html#QStringLiteral

fwcd replied:
Ah interesting, thanks!

Swiftb0y replied:
> But I'd wait for the MSVC build to complete first, I don't know whether cond-compiles are just problematic within parameter contexts or in expressions in general.

QStringLiteral is a Macro, so MSVC probably doesn't like preprocessor directives within the parameters of a macro. If QStringLiteral was just a constexpr function, it would probably be fine.

Swiftb0y replied:
Yes, thats even better.

Swiftb0y replied:
Though if we're really pedantic, I'd prefer to duplicate the semicolon, its rather odd on its own line with all that whitespace
```suggestion
static const QString kiTunesLocalhostToken =
#if defined(__WINDOWS__)
        QStringLiteral("//localhost/");
#else
        QStringLiteral("//localhost");
#endif
```

fwcd committed:
ITunesLocalhostToken: Use `QStringLiteral`

Co-authored-by: Swiftb0y <12380386+Swiftb0y@users.noreply.github.com>

fwcd committed:
ITunesMacOSImporter: Log NSError from ITLibrary initialization

Swiftb0y commented on src/library/itunes/itunesmacosimporter.mm:
```diff
@@ -0,0 +1,291 @@
+}
+
+ITunesImport ITunesMacOSImporter::importLibrary() {
+    @autoreleasepool {
```
could you explain `@autoreleasepool` to me? From what I could read so far I assume it causes all heap allocations within to be allocated in a memorypool which gets freed at the end of the block?

fwcd replied:
Yes, all Objective-C objects that are allocated within an `@autoreleasepool` block are released (i.e. deallocated) at the end of the block, as long as `autorelease` is called on them during initialization (e.g. `[[[Class alloc] init] autorelease]`.

After reading up on this a bit more, it turns out I was mistaken in that this is sufficient: Usually the compiler inserts this `autorelease` call automatically when ARC (Automatic Reference Counting) is enabled, but after experimenting around with it a bit it seems as if `clang` doesn't automatically enable ARC outside of Xcode projects, contrary to what I had assumed.

Therefore it looks like we have three options to avoid leaking memory:

- Manually insert `release`/`retain` calls (tedious)
- Call `autorelease` on every allocated ObjC object (slightly less tedious)
- Enable ARC in the `CMakeLists.txt` (lets the compiler auto-insert `release`/ `retain` calls)

The latter is the safest and most convenient solution, since we don't need to worry about ObjC memory management at all and can even remove the `@autoreleasepool` since the compiler will automatically make sure that objects are deallocated once the last reference goes out of scope. This is consistent with what Stackoverflow suggests:

- https://stackoverflow.com/questions/26241739/should-i-use-arc-when-coding-in-objective-c

For this reason, I've enabled ARC in e23b3f596df1e6533dfa96b154f1fd4296c04ae7 and removed the `@autoreleasepool`.

---

That `-fobjc-arc` is 'safe' can be verified with the following Objective-C program:

```objc
#import <Foundation/Foundation.h>

@interface Test : NSObject
- (void) dealloc;
@end

@implementation Test
- (void) dealloc {
  NSLog(@"Deallocing!");
}
@end

int main() {
  Test *test = [[Test alloc] init];
  return 0;
}
```

Compiling it with `clang -framework Foundation -fobjc-arc <file name>` causes the compiler to automatically release the object, thus printing `Deallocing`, whereas compiling it with `clang -framework Foundation <file name>` doesn't print anything (leaking the object).

Swiftb0y replied:
Thank you, I agree that adding automatic reference counting is best as long as we take care to avoid reference cycles.

fwcd commented on src/library/itunes/itunesfeature.h:
```diff
@@ -51,22 +54,23 @@ class ITunesFeature : public BaseExternalLibraryFeature {
     void clearTable(const QString& table_name);
     bool readNextStartElement(QXmlStreamReader& xml);
 
+    QString showOpenDialog();
```
That's the path to the selected file, but I agree, a comment would probably be helpful

fwcd commented on src/library/itunes/itunesmacosimporter.h:
```diff
@@ -0,0 +1,22 @@
+  public:
+    ITunesMacOSImporter(LibraryFeature* parentFeature,
+            QSqlDatabase& database,
+            std::atomic<bool>& cancelImport);
```
This is mostly an artifact from the existing iTunes XML importer which (even less safely) uses a non-atomic boolean for this. The `cancelImport` method + private member sounds good for now, I'd say.

fwcd replied:
Hm, digging into this a bit it might require a bit of a rearchitect. Currently imports are cancelled by setting `m_cancelImport = true` in `~ITunesFeature`. Translating this into a call (e.g. to a new method `ITunesImporter::cancelImport`) would require us to store the `ITunesImporter` in a field instead during `ITunesFeature::importLibrary` which doesn't feel cleaner than just passing the reference to the importer.

I agree that passing a reference to private members is fragile and requires us to be very careful to have the importer not outlive its 'parent' feature. Wdyt about making this a `std::shared_ptr<std::atomic<bool>>` instead of a reference?

Swiftb0y replied:
> would require us to store the ITunesImporter in a field instead during ITunesFeature::importLibrary which doesn't feel cleaner than just passing the reference to the importer.

Well I agree that storing the somewhat temporary `ITunesImporter` is not great but neither is the current implicit contract around that `m_cancelImport`. I think temporarily storing the pointer to the importer object is still an improvement.

> Then perhaps keeping it the current way and refactoring it in a future PR might be the best way to go about this, considering that it would also be deviation from the way this is currently handled in the Banshee/Traktor/Rhythmbox/etc. integration?


Yes, that definitely an option. Could you then revert those changes and instead leave a TODO summarizing what we've discussed so far? So primarily that we should add a **thread-safe** cancel function to the `ITunesImporter` interface (possibly to `BaseExternalLibraryFeature` as well) and that `m_cancelImport` needs to be made atomic everywhere.

fwcd replied:
The deviation I meant was w.r.t migrating the 'plain' boolean flag approach to a cancel method.

> Could you then revert those changes

Which ones specifically? 7d9ac9bd1d36d35d9b80f00ebc44e464a79aee8d? If we are going to use atomics in the future anyway, I don't see much harm in keeping it (from my understanding `bool&` is thread-unsafe, while passing `std::atomic<bool>&` is safe, but fragile/not very pretty due to passing references to private members around).

fwcd replied:
I've added some notes in 48250c10c4e2e065b4fd56b0ea6bddf6c50c161c as you suggested

fwcd committed:
Objective-C: Enable ARC and remove @autoreleasepool

This frees us from having to worry about Objective-C memory management
at all.

fwcd committed:
ITunesImporter: Remove unnecessary include

Co-authored-by: Swiftb0y <12380386+Swiftb0y@users.noreply.github.com>

fwcd committed:
ITunesFeature: Add short note on showOpenDialog

fwcd committed:
External libraries: Add todo notes regarding cancellation

uklotzde commented on src/library/itunes/itunesxmlimporter.h:
```diff
@@ -0,0 +1,40 @@
+    QString m_dbfile;
+    QSqlDatabase& m_database;
+    ITunesPathMapping& m_pathMapping;
+    std::atomic<bool>& m_cancelImport;
```
My internal, Rust-conditioned borrow checker is tilting if I see such a construct.

Wrapping the atomic into a shared_ptr would be much safer. The allocation only happens once during a session,

fwcd replied:
> My internal, Rust-conditioned borrow checker is tilting if I see such a construct.

Just imagine the importer being parameterized over `<'a>` and turn all of these into `&'a` 😄 

> Wrapping the atomic into a shared_ptr would be much safer

That what one solution I suggested in https://github.com/mixxxdj/mixxx/pull/11353#discussion_r1138980135 which was rejected though.

uklotzde replied:
@Swiftb0y I don't understand your comment on the reference count? This must always be atomic like in an `Arc` (C++ does not have `Rc`) and doesn't require additional wrapping.

Sure, I am aware that the importers are scoped somehow. But this scope or context is not modeled explicitly anywhere. I would expect at least a comment that explains these implicit assumptions.

Swiftb0y replied:
> I don't understand your comment on the reference count? This must always be atomic like in an Arc (C++ does not have Rc) and doesn't require additional wrapping.

Ah yes, you are correct, the shared_ptr docs on this are a bit hard to understand. Disregard my comment on `std::atomic<std::shared_ptr>` then. 

> Sure, I am aware that the importers are scoped somehow. But this scope or context is not modeled explicitly anywhere. I would expect at least a comment that explains these implicit assumptions.

Yes, a comment makes sense. I agree its not great, but its what we currently have. We already agreed-to solution of implementing virtual cancel member function solves the lifetime problem (in addition to problems I outlined when we had discussed them).

Swiftb0y replied:
Don't get me wrong, using a `std::shared_ptr` is fine for me. But throwing out the assumption that the `*Importer` doesn't outlive the `ITunesFeature` will cause even more problems.

fwcd replied:
So should we go with `shared_ptr` or keep it the current way?

uklotzde replied:
Add a comment to the member. For anyone who is not aware of the context when stumbling over this code.

uklotzde replied:
@Swiftb0y So using mutable references is now allowed? I remember that there was a _pointer-only policy_ for mutable references.

Swiftb0y replied:
True, I often forget that. we can make `m_cancelImport` a const reference. Or pointer, I don't care either way. Nonetheless, I struggle to understand why we are still discussing details of this sub-par design decision we already agreed to replace in a future PR.

uklotzde replied:
Using a const-ref would be sufficient here.

uklotzde replied:
The value is read-only for the subordinate components.

uklotzde replied:
The C++ atomic semantics differ from Rust. Without interior mutability atomics would not be useful in Rust.

fwcd replied:
I've updated these to `const` references where possible and added a small note on lifetimes in fb3cc2ab325e99fd78acffe3691d0403be8c0852

uklotzde commented on src/library/itunes/itunesmacosimporter.mm:
```diff
@@ -0,0 +1,288 @@
+        // interact well with Objective-C collections.
+
+        for (ITLibPlaylist* playlist in playlists) {
+            if (m_cancelImport) {
```
I recommend to use the explicit load operation from std::atomic instead of the implicit conversion operator.

fwcd committed:
ITunesImporter: Use const references where possible

fwcd committed:
ITunesImporter: Use explicit std::atomic<bool>::load

fwcd committed:
ITunesImporter: Add short comment on lifetime

Swiftb0y commented on src/library/itunes/itunesfeature.cpp:
```diff
@@ -402,51 +302,36 @@ TreeItem* ITunesFeature::importLibrary() {
+    ITunesImport iTunesImport;
+    std::unique_ptr<ITunesImporter> importer;
+
+#ifdef __MACOS_ITUNES_LIBRARY__
+    if (m_dbfile.isEmpty()) {
+        importer = std::make_unique<ITunesMacOSImporter>(this, m_database, m_cancelImport);
+    } else {
+        importer = std::make_unique<ITunesXMLImporter>(
+                this, m_dbfile, m_database, m_pathMapping, m_cancelImport);
```
I think we should log which importer was chosen here. Choosing the importer simply based on whether we found the db file seems somewhat brittle (though idk what else would be more robust).

fwcd replied:
The idea is that e.g. a user on macOS can explicit select an XML file, but also click `Use Default Library` in the context menu to switch back to the native iTunes integration. Perhaps we should make this more explicit?

I agree that logging the importer is probably a good idea.

Swiftb0y replied:
I don't quite understand the workflow, what context menu exactly? Is "native iTunes integration" the XML-file or the iTunesLibrary Framework solution?

fwcd replied:
> Is "native iTunes integration" the XML-file or the iTunesLibrary Framework solution?

The `iTunesLibrary` framework solution.

> I don't quite understand the workflow, what context menu exactly?

![image](https://user-images.githubusercontent.com/30873659/227251463-fb42291a-56c5-4351-b3a4-239f4b8e223a.png)

Selecting `Use Default Library` sets the dbfile path to an empty string which we interpret as "use the `iTunesLibrary` importer". If the dbfile path is set to an XML file, we use the XML importer instead. This should also handle existing users of the XML file integration on macOS gracefully (e.g. someone who has an iTunes XML on an external drive etc.)

Swiftb0y replied:
Mhmm, maybe add a comment that we interpret it that way idk. You're free to improve it in a future PR.

fwcd replied:
I've added a short doc comment to `m_dbfile` and a short log message

Swiftb0y reviewed (commented):
last comment. lgtm otherwise. I'll have to see whether I can conduct a manual test.

fwcd committed:
ITunesFeature: Clarify m_dbfile

fwcd committed:
ITunesMacOSImporter: Add doc comment

fwcd committed:
ITunesMacOSImporter: Skip DRM and remote tracks

fwcd committed:
ITunesXMLImporter: Add missing m_dbfile initialization

fwcd committed:
ITunesXMLImporter: Use better names for various members

fwcd committed:
ITunesXMLImporter: Use camelCase more consistently

fwcd commented:
I have made some minor stylistic improvements here and there and tested the XML importer more extensively, fixing a small regression that surfaced in the `guessMusicLibraryMountpoint` method (see 16b699e). Thus the path translation for relocated iTunes libraries should work correctly again.

Swiftb0y reviewed (commented):
small comment, personally, i think the changes look good, but I'd like some more testing, preferably by another 3rd party as I currently don't have access to apple hardware for testing.

fwcd commented on src/library/itunes/itunesfeature.cpp:
```diff
@@ -195,33 +170,37 @@ void ITunesFeature::activate(bool forceReload) {
         if (!dbSetting.isEmpty() && QFile::exists(dbSetting)) {
             m_dbfile = dbSetting;
         } else {
+#ifndef __MACOS_ITUNES_LIBRARY__
             // No Path in settings, try the default
+            // if not on macOS, where we use the ITunesMacOSImporter if m_dbfile is empty
             m_dbfile = getiTunesMusicPath();
+#endif
```
The `iTunesLibrary` framework is available on macOS 10.13 or later (see https://developer.apple.com/documentation/ituneslibrary), which is newer than the minimum version of Mixxx 2.3 (which runs on macOS 10.12 or later). If we wish to keep targeting 10.12 we could e.g. use an [`@available` check](https://developer.apple.com/documentation/swift/marking-api-availability-in-objective-c) to choose between the macOS importer and the XML importer at runtime.

fwcd replied:
I've factored out a `makeImporter` method and added an availability check to treat macOS 10.12 (or lower) like non-macOS platforms in 6e80d1609fc0894a431ddc7e18b9fd5e106113cf.

fwcd committed:
ITunesFeature: Factor out makeImporter and guard macOS version

Swiftb0y commented on src/library/itunes/itunesfeature.cpp:
```diff
@@ -8,55 +8,40 @@
+bool isMacOSImporterAvailable() {
+#ifdef __MACOS_ITUNES_LIBRARY__
+    // The iTunesLibrary framework is only available on macOS 10.13+
+    if (__builtin_available(macOS 10.13, *)) {
```
Can you add that as a comment for future reference?

fwcd replied:
Done

Swiftb0y commented:
@daschuer what do you think? Should we just merge for 2.4 and see what bug reports we get in the beta?

fwcd committed:
ITunesFeature: Comment on __builtin_available

daschuer commented:
I support merging this into 2.4. I think we need the confirmation that the legacy import is still working before merging.
Has this been tested?

daschuer reviewed (approved):
I have just fired up my win 10 virtual machine and can confirm that this does not creates obvious regression with iTunes 12.12.7.1

So lets merge this and collect bugs if there are any from beta testers.

Thank you for the work.

uklotzde commented:
> So lets merge this and collect bugs if there are any from beta testers.

That's the right attitude for making progress and keeping trustworthy developers motivated ;)

ronso0 commented on src/library/itunes/itunesfeature.h:
```diff
@@ -51,22 +55,31 @@ class ITunesFeature : public BaseExternalLibraryFeature {
     void clearTable(const QString& table_name);
```
asd
